### PR TITLE
Clean up and modularize build 

### PR DESCRIPTION
**Pull request by JessicaMeixner-NOAA**
# Pull Request Summary
This PR is a clean-up and modularization of the build. The expectation is that no answers will change. 

## Description
This PR adds a "build_utility.sh" function (copying the ideas from @arunchawla-NOAA) in preparation for cleaning up and modularizing the build before transition to CPP from switches and will help in adding additional build functionalities such as CMAKE.   Other updates include: 
* The READMEs in model/bin were consolidated to one README and additional build information such as  library dependency information was added there.  The READMES from various institutions were copied as is with the exception of formatting updates for readability. 
* The make_HYB make_MPI etc were deleted as they had not been maintained and were not 100% correct.  This will require fewer updates when moving to CPP from switches. 
* The "printer" and "C compiler" variables were removed from the wwatch3.env as the variables are not used. 
* Within the bin directory, the variable source was renamed to sterm as "source" can be a system variable. 
* The WWATCH3_NETCDF variable was removed as we only support 1 option and therefore it is no longer necessary. 

### Issue(s) addressed
* This is in prep for Issue #360 and is the last PR for the [code clean up project](https://github.com/NOAA-EMC/WW3/projects/1).  This will complete the [milestone](https://github.com/NOAA-EMC/WW3/milestone/26) for the code clean up as well. 

* Reviewers: @aliabdolali @mickaelaccensi 

### Commit Message
Clean up and modularize build

### Testing
* How were these changes tested? Regression tests
* Are the changes covered by regression tests? Yes. 
* If a new feature was added, was a new regression test added? No new features
* Have regression tests been run? yes
* Compiler / HPC used to run the regression tests in the PR: hera.intel 
* Output of matrix.comp:
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/7059740/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/7059741/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/7059742/matrixDiff.txt)

    
* Only the expected not b4b issues: 
```
mww3_test_03/./work_PR2_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UQ_MPI_d2                     (9 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (10 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (10 files differ)
mww3_test_03/./work_PR1_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (7 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (9 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR3_UQ_MPI_e_c                     (1 files differ)
mww3_test_07/./work_PR3_UQ                     (3 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (7 files differ)
ww3_tp2.17/./work_mc1                     (0 files differ)
ww3_ufs1.1/./work_b                     (0 files differ)
ww3_ufs1.2/./work_b                     (0 files differ)
ww3_ufs1.3/./work_a                     (1 files differ)
```

### CODE REVIEW

File: model/bin/make_makefile.sh [hunk 1]
```diff
@@ -91,712 +75,44 @@
...
-  fi
+# 3.c Create makefile header - - - - - - - - - - - - - - - - - - - - - - - - -
+
+  echo '# -------------------------'              > makefile
+  echo '# WAVEWATCH III makefile   '             >> makefile
+  echo '# -------------------------'             >> makefile
 
 
-# 2.c Make makefile and file list  - - - - - - - - - - - - - - - - - - - - - -
+# 3.d Go through programes to create file list - - - - - - - - - - - - - - -
```
Comment (mickaelaccensi) [inline model/bin/make_makefile.sh]: programs

File: model/bin/README.md [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    cd model/bin/ 
+    ./w3_make ww3_grid 
+
+
+# Below is information about various centers use of WW3: 
+
+## Information from Ifremer: 
+
+At Ifremer we run several wave models routinely using the WAVEWATCH codes, 
+these are part of the Previmer system and are reported under "SHM" or "SHOM"
```
Comment (mickaelaccensi) [inline model/bin/README.md]: MARC

Comment (mickaelaccensi) [reply model/bin/README.md]: LOPS

**Review by mickaelaccensi — COMMENTED** [review]
w3_automake does not work anymore, can you check ? also some minor corrections

**Comment by JessicaMeixner-NOAA** [comment]
> w3_automake does not work anymore, can you check ? also some minor corrections

Is w3_automake working on the top of the develop branch?  I'll try to look into why it's not working now.  I don't ever use it so I might have questions for you.  Thanks for looking and finding the other corrections!

File: model/bin/README.md (hunk 2 above)
Comment (JessicaMeixner-NOAA) [reply model/bin/README.md]: So this should say instead: 
these are part of the Previmer system and are reported under "MARC" or "LOPS"

Is that correct? @mickaelaccensi

**Comment by JessicaMeixner-NOAA** [comment]
@mickaelaccensi I think I figured out what is going on with w3_automake -- it's using sort_switches which was relying on a section of make_makefile which was moved to build_utils.  I'll push updates for this after a little more testing.  I'm also going to update the list of reg vs netcdf programs because it seems a little out of date with respect to w3_make.

**Comment by JessicaMeixner-NOAA** [comment]
@mickaelaccensi w3_automake should be working now, please let me know if you have any more issues.

File: model/bin/README.md (hunk 2 above)
Comment (mickaelaccensi) [reply model/bin/README.md]: these are part of the MARC system and are reported under "LOPS"

**Comment by mickaelaccensi** [comment]
it works now, I'll do the regtests and keep you in touch

File: model/bin/README.md (hunk 2 above)
Comment (JessicaMeixner-NOAA) [reply model/bin/README.md]: Updated -- thanks for the clarification.

**Comment by mickaelaccensi** [comment]
regtests are ok except for ww3_tp2.17 which attempts to run ww3_grib by default, so I had to disable the grib output.
another error is in ww3_tr1 :
ww3_tr1/work_bndin/ww3_bound.out:IP, file, I/O error:    1, ../input_bndin/bnd1.spc,     0
ww3_tr1/work_bndin/ww3_bound.out:IP, file, I/O error:    2, ../input_bndin/bnd2.spc,     0
ww3_tr1/work_bndin/ww3_bound.out:IP, file, I/O error:    3, ../input_bndin/bnd3.spc,     0
but this one is also present in the develop branch, I'll raise an issue for this one

**Comment by JessicaMeixner-NOAA** [comment]
@mickaelaccensi the ww3_tp2.17 should be an issue for you in the develop branch too? Or is it just in this branch you have this issue?

**Comment by mickaelaccensi** [comment]
it is in both develop and feature branch.

**Comment by JessicaMeixner-NOAA** [comment]
> it is in both develop and feature branch.

Since it's in develop as well, I do not plan to address this in this PR then.  Is that okay @mickaelaccensi

**Comment by mickaelaccensi** [comment]
ok

File: manual/impl/compile.tex [hunk 3]
```diff
@@ -2,20 +2,18 @@
 \subsection{~Optional environment settings}
 \vssub
 
-Compilation of the \ws\ NetCDF enabled programs requires the environment variable
-{\code WWATCH3\_NETCDF} be set to {\code NC4} (compile with NetCDF version 4.x).  
-{\code NC3} is no longer supported. 
-If {\code WWATCH3\_NETCDF = NC4}, then the following environment variable
-is required.
+Compilation of the \ws\ NetCDF enabled programs requires NetCDF version 4.1.1 or higher.
+The following enviroment variable is also required:
```
Comment (ukmo-ccbunney) [inline manual/impl/compile.tex]: typo: enviroment => environment

File: model/bin/README.md [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+## Optional variable to use with any configuration: 
+
+WW3_PARCOMPN = <how many parallel make tasks to use, default to 4 if not set >
+
+
+## To build NetCDF executables you need:
+
+The requirements for NetCDF are: 
+* NetCDF version 4.1.1 or higher (Use "nc-config --version" to check) 
+* NetCDF-4 API enableed (Use "nc-config --has-nc4" to check)
```
Comment (ukmo-ccbunney) [inline model/bin/README.md]: typo: enableed => enabled.

**Review by ukmo-ccbunney — CHANGES_REQUESTED** [review]
Changes look good to me.
Just a few typos.
I am running the regression test and will post the results later.

File: manual/impl/compile.tex (hunk 3 above)
Comment (JessicaMeixner-NOAA) [reply manual/impl/compile.tex]: Fixed!

File: model/bin/README.md (hunk 4 above)
Comment (JessicaMeixner-NOAA) [reply model/bin/README.md]: fixed!

**Comment by JessicaMeixner-NOAA** [comment]
Thanks for the typo corrections @ukmo-ccbunney and looking at everything, let me know if you run into any issues with the regtests.

**Review by ukmo-ccbunney — APPROVED** [review]
@JessicaMeixner-NOAA
Results from the regression tests on our HPC. All looks ok, apart from the usual mww3_test_03 differences and a couple of expected differences explained below:
```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR2_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (10 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (11 files differ)
mww3_test_03/./work_PR1_MPI_e                     (1 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (13 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (12 files differ)
mww3_test_03/./work_PR2_UQ_MPI_d2                     (11 files differ)
ww3_tp2.17/./work_mc1                     (1 files differ)
ww3_tp2.17/./work_ma1                     (1 files differ)
ww3_ts1/./work_NL5                     (3 files differ)
```

**ww3_tp2.17** is currently not working for me at all in develop due to some input file issues.

**ww3_ts1/work_NL5** has been giving problems with other PRs (see issue #471 I just raised). However, when I re-ran that test, it came back B4B.

**Comment by JessicaMeixner-NOAA** [comment]
@mickaelaccensi you have not officially approved this PR, are there any other issues you'd like me to address related to this PR?  

@thesser1 any feedback on your end?

**Comment by JessicaMeixner-NOAA** [comment]
All WW3 regtests are still passing: 

[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/7154304/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/7154305/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/7154306/matrixDiff.txt)

ufs-weather-model wave tests on hera also passed

**Review by aliabdolali — APPROVED** [review]

**Comment by aliabdolali** [comment]
I ran the tests and they all passed with pre-known non-b4b cases.

**Pull request merged**